### PR TITLE
Network Exception Handling Transactions

### PR DESCRIPTION
**Pull request by annym**
- Adding support to NetworkException handling.
- Wrapping NetworkExceptions into TransactionAbortedExceptions when in
  the context of a transaction.
- Integration Tests for simple transactions and nested
  transactions.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **772e25c**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #618 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/NKB5OrOe3dfHxp8fzV3kflanbxij22D8)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **85c2c8f**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #618 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/ITWEyk86nbVbl3aCD95YrlJMvnhPi7Vt)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0700a54**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #618 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/Vfqlg2usj7RURTNQc3qpze1v5VLKJtZO)

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java [hunk 1]
```diff
@@ -204,8 +220,13 @@ public void sync() {
         // If we aren't coming from a transactional context,
         // redirect us to a transactional context first.
         if (TransactionalContext.isInTransaction()) {
-            return (R) TransactionalContext.getCurrentContext()
-                    .getUpcallResult(this, timestamp, conflictObject);
+            try {
+                return (R) TransactionalContext.getCurrentContext()
+                        .getUpcallResult(this, timestamp, conflictObject);
+            } catch (Exception e) {
+                log.debug("UpcallResultInner[{}] Exception: {}", this, e);
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: Just call this UpcallResult[{}]

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java [hunk 2]
```diff
@@ -281,6 +313,9 @@ public UUID getStreamID() {
                 catch (Exception ex) {}
                 sleepTime = min(sleepTime * 2L, maxSleepTime);
                 retries++;
+            } catch (Exception e) {
+                log.debug("TXExecute[{}] Exception: {}", this, e);
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: Add that we are aborting this exception (TXExecute[{}] Abort with Exception: {})

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java (hunk 1 above)
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: Done

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java (hunk 2 above)
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: Done

**Comment by Maithem** [comment]
@annym you can do : 
git rebase -i HEAD~20
# then re-arrange the commits such that the last 4 commits look like this:
pick 772e25c Network Exception Handling Transactions
pick 4136ce4 Fix to Corfu Server Shutdown Method
pick 85c2c8f Fixing Integration Test Bug
pick ccd6d8a Change on log info message

# complete the rebase and you'll have to resolve some conflicts in:
	both modified:      runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java
	both modified:      runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java
	both modified:      runtime/src/main/java/org/corfudb/runtime/view/ObjectsView.java
	both modified:      test/src/test/java/org/corfudb/runtime/view/ObjectsViewTest.java


# after that you can do a trivial squash of consecutive commits

**Comment by annym** [comment]
Is this ready to be merged into master?

**Pull request merged**